New WebSockets client guide (#40931)
mdn/content: files/en-us/web/api/websockets_api/writing_websocket_client_applications/index.md

page-type: guide
---

{{DefaultAPISidebar("WebSockets API")}}

In this guide we'll walk through the implementation of a WebSocket-based ping application. In this application, the client sends a "ping" message to the server every second, and the server responds with a "pong" message. The client listens for "pong" messages and logs them, keeping track of how many message exchanges there have been.

Although this is a pretty minimal application, it covers the fundamental points involved in writing a WebSocket client.

You can find the complete example at [https://github.com/mdn/dom-examples/tree/main/websockets](https://github.com/mdn/dom-examples/tree/main/websockets). The server side is written in [Deno](https://deno.com/), so you'll have to install that first if you want to try running the example locally.

## Creating a `WebSocket` object

In order to communicate using the WebSocket protocol, you need to create a {{domxref("WebSocket")}} object. As soon as you create this object, it will start trying to connect to the server.

```js
const wsUri = "ws://127.0.0.1/";
const websocket = new WebSocket(wsUri);
```

> The `WebSocket` constructor takes one mandatory argument, which is the URL of the WebSocket server to connect to. In this case, since we're running the server locally, we're using the address of localhost.

Author: Added in https://github.com/mdn/content/pull/40931/commits/9e4076fdddb7af3d145df010490defb352ecafbc. I don't think this note is great, and I don't fully understand how protocols work with WebSockets - this is a thing that ought to be explained in our docs, but I think it is out of scope for my change here.

> The `WebSocket` constructor takes one mandatory argument, which is the URL of the WebSocket server to connect to. In this case, since we're running the server locally, we're using the address of localhost.

Reviewer: FWIW I think the note is good!

```js
websocket.addEventListener("open", () => {
  log("CONNECTED");
  pingInterval = setInterval(() => {
    log(`SENT: ping: ${counter}`);
    websocket.send("ping");
  }, 1000);
});
```

## Listening for errors

If an error occurs while the connection is being established or at any time after it is established, the {{domxref("WebSocket.error_event", "error")}} event will be fired.

Our application doesn't do anything special on error, but we do log it:

```js
websocket.addEventListener("error", (e) => {
  log(`ERROR`);
});
```

On an error, the connection is closed and the `close` event will be fired.

## Sending messages

We've already seen that once the connection is established, we can use the {{domxref("WebSocket.send()", "send()")}} method to send messages to the server:

```js
websocket.addEventListener("open", () => {
  log("CONNECTED");
  pingInterval = setInterval(() => {
    log(`SENT: ping: ${counter}`);
    websocket.send("ping");
  }, 1000);
});
```

In our example we send text, but you can also send binary data as a {{domxref("Blob")}}, {{jsxref("ArrayBuffer")}}, {{jsxref("TypedArray")}}, or {{jsxref("DataView")}}.

> A common approach is to use {{glossary("JSON")}} to send and receive serialized JavaScript objects. For example, instead of just sending the text message "ping", our client could send a serialized object including the number of messages exchanged so far:

Reviewer 2: ```suggestion
A common approach is to use {{glossary("JSON")}} to send serialized JavaScript objects as text. For example, instead of just sending the text message "ping", our client could send a serialized object including the number of messages exchanged so far:
```

```js
const message = {
  iteration: counter,
  content: "ping",
};
websocket.send(JSON.stringify(message));
```

> The `send()` method is asynchronous, and the {{domxref("WebSocket.bufferedAmount")}} property represents the number of bytes that have not yet been transmitted. Note that the WebSockets protocol uses {{glossary("UTF-8")}} to encode text, so `bufferedAmount` is calculated based on the UTF-8 encoding of any buffered text data.

Author: To me it is async because the caller doesn't have to wait for the data to be sent before the function returns and the caller can continue:

```js
s.send(data);
// the data isn't actually sent when we get here
```

I appreciate it is unusual in that there is no way the caller is told that the operation has completed (i.e. with a promise or a callback or an event). But I would still call it async. Still I don't think getting hung up on the terminilogy is very useful so I'm happy to remove the word.

```suggestion
The `send()` method does not wait for the data to be transmitted before returning to the caller. It just adds the data to the queue and begins the process of transmission. The {{domxref("WebSocket.bufferedAmount")}} property represents the number of bytes that have not yet been transmitted. Note that the WebSockets protocol uses {{glossary("UTF-8")}} to encode text, so `bufferedAmount` is calculated based on the UTF-8 encoding of any buffered text data.
```

> Do you want to dig into bufferedAmount a bit more. Is it specifically the amount of bytes in the queue that are as yet unsent?

Isn't that just what "The {{domxref("WebSocket.bufferedAmount")}} property represents the number of bytes that have not yet been transmitted" says?

> The `send()` method is asynchronous, and the {{domxref("WebSocket.bufferedAmount")}} property represents the number of bytes that have not yet been transmitted. Note that the WebSockets protocol uses {{glossary("UTF-8")}} to encode text, so `bufferedAmount` is calculated based on the UTF-8 encoding of any buffered text data.

Reviewer: FWIW I agree it's async. A promise or callback isn't necessary for a call to be considered async (although it's nice when it does!)

## Receiving messages

To receive messages from the server, we listen for the {{domxref("WebSocket.message_event", "message")}} event.

Our message event handler logs the received message, and increments our count of the number of message exchanges that have occurred:

```js
websocket.addEventListener("message", (e) => {
  log(`RECEIVED: ${e.data}: ${counter}`);
  counter++;
});
```

The server can also send binary data, which is exposed to clients as a {{domxref("Blob")}} or an {{jsxref("ArrayBuffer")}}, based on the value of the {{domxref("WebSocket.binaryType")}} property.

> As we saw for sending messages, the server could also send JSON-serialized objects, which the client can then parse:

Reviewer 2: ```suggestion
As we saw for sending messages, the server can also send JSON strings, which the client can then parse into an object:
```

```js
websocket.addEventListener("message", (e) => {
  const message = JSON.parse(e.data);
  log(`RECEIVED: ${message.iteration}: ${message.content}`);
  counter++;
});
```

## Handling disconnect

> When the connection is closed, because either client or server closed it or because there was an error, then the {{domxref("WebSocket.close_event", "close")}} event will be fired.

Reviewer 2: ```suggestion
When the connection is closed because either the client or the server closed it or an error occurred, the {{domxref("WebSocket.close_event", "close")}} event will be fired.
```

Author: ```suggestion
When the connection is closed, because either the client or the server closed it or because an error occurred, the {{domxref("WebSocket.close_event", "close")}} event will be fired.
```

I think we need the extra "because" or it is ambiguous:

- the client closed it
- the server closed it or an error occurred

rather than:

- the client closed it
- the server closed it
- an error occurred

Our application listens for the `close` event and cleans up the interval timer when it is fired:

```js
websocket.addEventListener("close", () => {
  log("DISCONNECTED");
  clearInterval(pingInterval);
});
```

## Working with the bfcache

The back/forward cache, or {{glossary("bfcache")}}, enables much faster back and forward navigation between pages that the user has recently visited. It does this by storing a complete snapshot of the page, including the JavaScript heap.

> The browser pauses and then resumes JavaScript execution when a page is added to or loaded from the bfcache. This means that, depending on what the page is doing, it's not always safe for the browser to use the bfcache for the page. If the browser determines that it is not safe, then the page will not be added to the bfcache, and then the user will not get the performance benefit that it can bring.

Reviewer 2: ```suggestion
The browser pauses and then resumes JavaScript execution when a page is added to or loaded from the bfcache. This means that, depending on what the page is doing, it's not always safe for the browser to use the bfcache for the page. If the browser determines that it is not safe, the page will not be added to the bfcache, and the user will not get the performance benefit that it can bring.
```

Reviewer 2: Also, maybe link to https://developer.mozilla.org/en-US/docs/Web/API/Performance_API/Monitoring_bfcache_blocking_reasons for more information on why pages might not be added to the bfcache.

Author: I think it's a shame that this is presented as a page on how to use a Chrome-only feature. It would be better to have a page under https://developer.mozilla.org/en-US/docs/Web/Performance, that talks about the bfcache as it works across browsers, and the various reasons pages might not be bfcache-friendly in various browsers, and links to `notRestoredReasons` as a tool you can use in that context.

Author: Yes, sorry, I didn't mean to sound so negative about it! It is a good article and obviously a useful API. And we should link to it. I just think it would be great to have a browser-agnostic article about bfcache compatibility on MDN, that's not tied to specific APIs.

Author: -> https://github.com/mdn/content/pull/40931/commits/ef5e49c14e077ebfc37950ec7aeb4078cbd6bdca and https://github.com/mdn/content/pull/40931/commits/86ebca5c5d6e506d43b0dd6b39fbe914ab605dd6

Different browsers use different criteria for adding a page to the bfcache. If a page has an active WebSocket connection, then:

- Firefox and Chrome will not add the page to the bfcache.

> - Safari will add the page to the bfcache, but will cancel any active network requests. This is likely to generate an error in your WebSocket connection, leading to the connection being closed.

Reviewer: LGTM!

In Chrome, you can make a WebSocket client compatible with the bfcache by closing your WebSocket connection in the handler for the {{domxref("Window.pagehide_event", "pagehide")}} event, which is fired before the page is added to the bfcache.

We do this in our example app:

```js
window.addEventListener("pagehide", () => {
  if (websocket) {
    log("CLOSING");
    websocket.close();
    websocket = null;
    window.clearInterval(pingInterval);
  }
});
```

> Conversely, by listening for the {{domxref("Window.pageshow_event", "pageshow")}} event, you can seamlessly start the connection again when the page is loaded, or when it is retrieved from the bfcache. So in our example we will add all the code to initialize our WebSocket and set up its event listeners inside the `pageshow` event handler:

Reviewer 2: ```suggestion
Conversely, by listening for the {{domxref("Window.pageshow_event", "pageshow")}} event, you can seamlessly start the connection again when the page is loaded or retrieved from the bfcache. In our example, we add all the code to initialize our WebSocket and set up its event listeners inside the `pageshow` event handler:
```

Author: Again I think I like my extra words around ", or when it is". I think it makes the structure clearer. I suppose this is subjective?

Author: > pageshow is also fired on page load. So might wanna mention that (to avoid people opening a connection twice on pageload).

I do mention this:

> you can seamlessly start the connection again when the page **is loaded or** retrieved from the bfcache.

good enough?

let websocket = null;

> window.addEventListener("pageshow", () => {

Reviewer 3: Great catch, you are right.

I think that this demo is using the `pageshow` event handler to both initialize the WebSocket connection at initial load as well as restoring the same logic for BFCache restore.

I believe most developers will not use the pageshow event handler this way.

+1 to change this to initialize WebSocket normally on pageload and run the same logic on pageshow event.persisted=true to make it work again for BFCache.

> window.addEventListener("pageshow", () => {

Reviewer 4: Thanks both for the feedback. I've created an issue to track this here so that this doesn't get lost in the chatter:

- [ ] https://github.com/mdn/content/issues/41357

Feel free to add more context or details in the issue 👍🏻

> window.addEventListener("pageshow", () => {

Author: > I think that this demo is using the pageshow event handler to both initialize the WebSocket connection at initial load as well as restoring the same logic for BFCache restore.
>
> I believe most developers will not use the pageshow event handler this way.

For my education, why not?

window.addEventListener("pageshow", () => {
  log("OPENING");

  websocket = new WebSocket(wsUri);

  websocket.addEventListener("open", () => {
    log("CONNECTED");
    pingInterval = setInterval(() => {
      log(`SENT: ping: ${counter}`);
      websocket.send("ping");
    }, 1000);
  });

  websocket.addEventListener("close", () => {
    log("DISCONNECTED");
    clearInterval(pingInterval);
  });

  websocket.addEventListener("message", (e) => {
    log(`RECEIVED: ${e.data}: ${counter}`);
    counter++;
  });

  websocket.addEventListener("error", (e) => {
    log(`ERROR: ${e.data}`);
  });
});
```

> If you run our example, try navigating to a different page, then back to the example. In Chrome, you should see that the example starts the connection again, and keeps its original context: so, for example, it remembers the count of exchanged messages.

Reviewer 2: In Firefox and Safari, it just seems to start the app again from its initial state, which is a shame. Worth mentioning?

Author: This is not what I see in Safari, and I would like to know why we see different things here.

But it is what I see in Firefox, and that's expected: because the page isn't added to the bfcache, then it is reloaded, so the log starts out empty and the counter starts again at zero.

In Safari, after a forward/back navigation, I see this:

<img width="271" height="642" alt="Screenshot 2025-09-02 at 8 57 24 AM" src="https://github.com/user-attachments/assets/8635be9f-1b1b-4719-a120-e1456ce41d2a" />

Here, I clicked "forward" after exchange 4, and then "back". I think what happens here is that:

- the page was added to/loaded from the bfcache, so the HTML/JS is not reloaded, and we keep the contents of the lof and the counter value
- but because Safari closed the network request as soon as we navigated forward, we get a WebSocket error notification

It is perhaps worth spelling this out in the guide, maybe not to the screenshot level of detail though.

Reviewer 3: Hi @wbamberg, thanks for raising this.

I would love to hear from someone from WebKit if possible if they can share what Safari is doing with WebSocket.

I also noticed that Safari is behaving somehow differently from Chrome and Firefox.

It looks like pages using WebSocket without closing the connection via pagehide are still eligible for BFCache on Safari somehow.
From my tests it looked like the approach of closing the connection via pagehide and restarting the connection via pageshow was also working in Safari.

Can you please share the error you are getting?

Reviewer 2: > This is not what I see in Safari, and I would like to know why we see different things here.

Ah, hold on. I've tried it again in Safari. When I navigate to a new page by typing the URL into the address bar and pressing Enter, then press "Back", I get the behavior I describe (I am using gamespot.com as a test).

When I then press "Forward" to go back to the new page, then press "Back", I get the behavior your screenshot is showing.

I've tried it with a few other example pages, and get the same effects,

Author: Yes, I see that as well. In https://github.com/mdn/content/pull/40931/commits/1b0a04944397070b1bf1db4922b58d79d0c8e03d I've been less specific about what the browsers do, so I hope that means we don't have to worry about this.
